redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt

```diff
@@ -28,128 +28,284 @@ import app.cash.redwood.lazylayout.widget.RefreshableLazyList
+      val windowedIndex: Int = max(index - itemsBefore, 0)
+
+      return if (windowedIndex < viewPortList.count()) {
+        viewPortList[windowedIndex]
+      } else {
+        println("🔴 Error: index: $windowedIndex is outside of ${viewPortList.count()} item window size")
+
+          // TODO: Replace with an error item
+        viewPortList[0]
+      }
+    }
+
+    override fun itemCount(): Int {
+      return max(itemsBefore - 1, 0 ) + viewPortList.count() + itemsAfter
     }
   }
 
-  private val tableViewDelegate: UITableViewDelegateProtocol =
-    object : NSObject(), UITableViewDelegateProtocol {
-      override fun tableView(
-        tableView: UITableView,
-        willDisplayCell: UITableViewCell,
-        forRowAtIndexPath: NSIndexPath,
-      ) {
-        val content = itemsList[forRowAtIndexPath.item.toInt()]
-        (willDisplayCell as Cell).setView(content.value)
+  private val viewPortListCoordinator = object {
+    var minIndex: Int = 0
+    var maxIndex: Int = 0
+
+    fun notifyViewportChanged() {
+      onViewportChanged(minIndex, maxIndex)
+    }
+
+    fun updateViewport(minIndex: Int, maxIndex: Int) {
+      if (minIndex != this.minIndex || maxIndex != this.maxIndex) {
+        this.minIndex = minIndex
+        this.maxIndex = maxIndex
+
+        println("🔵 viewport: [$minIndex $maxIndex]")
+        notifyViewportChanged()
       }
     }
+  }
+  // A callback to tell LazyList that we have an update to the window of items available
+  private lateinit var onViewportChanged: (firstVisibleItemIndex: Int, lastVisibleItemIndex: Int) -> Unit
+
+  // UICollectionView + Protocols
+
+  // Must be initialized before the collectionView
+  private var collectionViewFlowLayout: UICollectionViewFlowLayout =
+    object: UICollectionViewFlowLayout() {}
+
+  private val collectionViewDataSource: UICollectionViewDataSourceProtocol =
+    object : NSObject(), UICollectionViewDataSourceProtocol {
+
+      override fun collectionView(
+        collectionView: UICollectionView,
+        numberOfItemsInSection: NSInteger,
+      ): NSInteger {
+        return items.itemCount().toLong()
+      }
 
-  private val tableViewDataSource: UITableViewDataSourceProtocol =
-    object : NSObject(), UITableViewDataSourceProtocol {
-      override fun numberOfSectionsInTableView(
-        tableView: UITableView,
-      ): NSInteger = 1L
+      override fun collectionView(
+        collectionView: UICollectionView,
+        cellForItemAtIndexPath: NSIndexPath,
+      ): UICollectionViewCell {
+        val cell = collectionView.dequeueReusableCellWithReuseIdentifier(
+          identifier = reuseIdentifier,
+          forIndexPath = cellForItemAtIndexPath
+        ) as LazyListContainerCell
 
-      override fun tableView(
-        tableView: UITableView,
-        numberOfRowsInSection: NSInteger,
-      ): NSInteger = itemsList.size.toLong()
+        val widget = items.itemForGlobalIndex(cellForItemAtIndexPath.item.toInt())
```

Comment:
@veyndan, what if instead of having to manage a list on this end of things. we just have a reference that can be used here and then have the viewport and recycling logic happen on the Redwood LazyList side? This way we can save ourselves the several round trips to make sure we have the item

Reply:
I'm not totally sure what you mean by:

> we just have a reference that can be used here

Reply:
i meant migrating over the `Items` object to be handled by redwood directly so that we can just give the LazyList implementations a callback for 

`itemAtIndex(index: Index) -> Item` and not have to do state management here. but this probably comes with a performance cost?

Reply:
The logic would have to continue to live on the host side, otherwise having `Items` on the guest side would involve  a lot of Zipline serialization — something which we had before and was causing a lot of performance headaches.

Perhaps we could share the implementation across Android View's though? `ViewLazyList` [has a similar-ish concept](https://github.com/cashapp/redwood/blob/e3b9bc927b1496c02264dcafce4f9e6e68b7d6e3/redwood-lazylayout-view/src/main/kotlin/app/cash/redwood/lazylayout/view/ViewLazyList.kt#L66-L94) to what you've got here. Can probably be a follow-up though!